clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -415,6 +439,27 @@ public ConsumerRecords<K, V> poll(final Duration timeout) {
             return ConsumerRecords.empty();
         } finally {
             kafkaConsumerMetrics.recordPollEnd(timer.currentTimeMs());
+            wakeupTrigger.clearActiveTask();
+        }
+    }
+
+    private CompletableFuture<Void> setupWakeupTrigger() {
+        final CompletableFuture<Void> wakeupFuture = new CompletableFuture<>();
+        return wakeupTrigger.setActiveTask(wakeupFuture);
```

Comment:
I had to scratch my head a little bit over this.

Since we already have the `Wakeupable` interface, I wonder if it wouldn't be cleaner to have a new static subclass:

```
/** Placeholder for a task that delays wake-up until it's manually triggered */
DelayedWakeupTask(final boolean wakeUpMarker)
```

Then replace `DelayedWakeupTask(false)` by `DelayedWakeupTask(true)` in `WakeupTrigger.wakeup` and move `maybeTriggerWakeup` to `WakeupTrigger`, rename it to `maybeTriggerDelayedWakeup`.

Then we'd have fewer code paths to worry about (no non-exceptional completion, no other possible exceptions where we ourselves don't know what to do and just throw and `IllegalStateException`).

Its just an idea, could be that I overlooked something.

Reply:
I changed it to not use a future, but a little bit different from you proposal. The code became indeed simpler.